### PR TITLE
check-in ModelConfig

### PR DESCRIPTION
**Pull request by AllentDan**

### CODE REVIEW

File: lmdeploy/model.py [hunk 1]
```diff
@@ -18,6 +18,31 @@ class SamplingParam:
     repetition_penalty: float = 1.0
 
 
+@dataclasses.dataclass
+class ModelConfig:
+    """parameters for chat templates."""
+    model_name: str  # determine which chat template will be applied
+    system: Optional[str] = None  # begin of the system prompt
```
Comment (lvhan028) [inline lmdeploy/model.py:25]: Shouldn't the default values of the attributes of string type be empty strings?

Comment (AllentDan) [reply lmdeploy/model.py:25]: The default value should be our settings in model.py. `None` is a flag here to determine whether we should modify default settings.

Comment (AllentDan) [reply lmdeploy/model.py:25]: Suppose the user wants to disable our default settings. He will set `meta_instructions=''`.

**Comment by lvhan028** [comment]
How `ModelConfig` can help improve the user experience? Could you please provide a piece of pseudocode?

**Comment by AllentDan** [comment]
> How `ModelConfig` can help improve the user experience? Could you please provide a piece of pseudocode?

```python
from lmdeploy import pipeline, GenerationConfig​, ModelConfig
from lmdeploy.turbomind import EngineConfig​
pipe = lmdeploy.pipeline('internlm/internlm-chat-7b', backend_config=EngineConfig())​
gen_config = GenerationConfig(max_new_tokens=100, top_p=0.8, top_k=40, temperature=0.8)​
model_config = ModelConfig(meta_instruction='[fake meta_instructin]')​
response = pipe(prompts, gen_config, model_config)
```

File: lmdeploy/model.py [hunk 2]
```diff
@@ -18,6 +18,31 @@ class SamplingParam:
...
+    eosys: Optional[str] = None  # end of the system prompt
+    user: Optional[str] = None  # begin of the user prompt
+    eoh: Optional[str] = None  # end of the user prompt
+    assistant: Optional[str] = None  # begin of the assistant prompt
+    eoa: Optional[str] = None  # end of the assistant prompt
+    capability: Optional[str] = None
+
+    @property
+    def model(self):
+        attrs_dict = {
```
Comment (grimoire) [inline lmdeploy/model.py:36]: dataclasses.asdict

File: lmdeploy/model.py [hunk 3]
```diff
@@ -18,6 +18,31 @@ class SamplingParam:
...
+    meta_instruction: Optional[str] = None  # system prompt
+    eosys: Optional[str] = None  # end of the system prompt
+    user: Optional[str] = None  # begin of the user prompt
+    eoh: Optional[str] = None  # end of the user prompt
+    assistant: Optional[str] = None  # begin of the assistant prompt
+    eoa: Optional[str] = None  # end of the assistant prompt
+    capability: Optional[str] = None
+
+    @property
+    def model(self):
```
Comment (grimoire) [inline lmdeploy/model.py:35]: Setting builder method as property is a little bit strange.

**Comment by lvhan028** [comment]
May remove `def main(...)` in `model.py`

File: lmdeploy/model.py (hunk 3 above)
Comment (AllentDan) [reply lmdeploy/model.py:35]: This is not supposed to be called frequently. And the build process is not computationally intensive.

**Review by lvhan028 — APPROVED** [review]

**Review by grimoire — APPROVED** [review]

**Pull request merged**